src/platform/linux/x11/mod.rs

```diff
@@ -142,25 +135,40 @@ impl EventsLoop {
                 (xlib.XNextEvent)(self.display.display, &mut xev);
             }
             self.process_event(&mut xev, &mut callback);
-            if self.interrupted.load(::std::sync::atomic::Ordering::Relaxed) {
-                break;
-            }
         }
     }
 
-    pub fn run_forever<F>(&self, mut callback: F)
-        where F: FnMut(Event)
+    pub fn run_forever<F>(&mut self, mut callback: F)
+        where F: FnMut(Event) -> ControlFlow
     {
-        self.interrupted.store(false, ::std::sync::atomic::Ordering::Relaxed);
+        self.pending_wakeup.store(false, atomic::Ordering::Relaxed);
 
         let xlib = &self.display.xlib;
 
         let mut xev = unsafe { mem::uninitialized() };
 
         loop {
             unsafe { (xlib.XNextEvent)(self.display.display, &mut xev) }; // Blocks as necessary
-            self.process_event(&mut xev, &mut callback);
-            if self.interrupted.load(::std::sync::atomic::Ordering::Relaxed) {
+
+            let mut control_flow = ControlFlow::Continue;
+
+            if self.pending_wakeup.load(atomic::Ordering::Relaxed) {
```

Comment:
The `pending_wakeup` flag is only an indicator of whether or not `wakeup` has been called on an `EventsLoopProxy`. It is `EventsLoopProxy::wakeup` itself that should [cause the next event function to return](https://github.com/tomaka/winit/pull/191/files#diff-c3f6b02da9e467b737e84b8236fa5ba2R562), though I could certainly be wrong on this - IIRC, I took this code from the old `WindowProxy::wakeup` method.

If you're on x11, you can test if this is working or not by running the included `proxy.rs` example. If it needs fiixing, please feel free to submit a PR to my fork, or let me know how I can go about fixing it.